trunk/docs/nikto.1
.\"     Title: nikto
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: 01/19/2010
.\"    Manual: 
.\"    Source: 
.\"
.TH "NIKTO" "1" "01/19/2010" "" ""
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
nikto \- Scan web server for known vulnerabilities
.SH "SYNOPSIS"
.HP 21
\fB/usr/local/bin/nikto\fR [options...]
.SH "DESCRIPTION"
.PP
Examine a web server to find potential problems and security vulnerabilities, including:
.sp
.RS 4
\h'-04'\(bu\h'+03'Server and software misconfigurations
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Default files and programs
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Insecure files and programs
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Outdated servers and programs
.RE
.PP
Nikto is built on LibWhisker (by RFP) and can run on any platform which has a PERL environment\&. It supports SSL, proxies, host authentication, IDS evasion and more\&. It can be updated automatically from the command\-line, and supports the optional submission of updated version data back to the maintainers\&.
.SH "OPTIONS"
.PP
Below are all of the Nikto command line options and explanations\&. A brief version of this text is available by running Nikto with the \-h (\-help) option\&.
.PP
\fB\-Cgidirs\fR
.RS 4
Scan these CGI directories\&. Special words "none" or "all" may be used to scan all CGI directories or none, (respectively)\&. A literal value for a CGI directory such as "/cgi\-test/" may be specified (must include trailing slash)\&. If this is option is not specified, all CGI directories listed in config\&.txt will be tested\&.
.RE
.PP
\fB\-config\fR
.RS 4
Specify an alternative config file to use instead of the config\&.txt located in the install directory\&.
.RE
.PP
\fB\-dbcheck\fR
.RS 4
Check the scan databases for syntax errors\&.
.RE
.PP
\fB\-Display\fR
.RS 4
Control the output that Nikto shows\&. See Chapter 5 for detailed information on these options\&. Use the reference number or letter to specify the type, multiple may be used:
.sp
1 \- Show redirects
.sp
2 \- Show cookies received
.sp
3 \- Show all 200/OK responses
.sp
4 \- Show URLs which require authentication
.sp
D \- Debug Output
.sp
V \- Verbose Output
.RE
.PP
\fB\-evasion\fR
.RS 4
Specify the LibWhisker IDS evasion technique to use (see the LibWhisker docs for detailed information on these)\&. Use the reference number to specify the type, multiple may be used:
.sp
1 \- Random URI encoding (non\-UTF8)
.sp
2 \- Directory self\-reference (/\&./)
.sp
3 \- Premature URL ending
.sp
4 \- Prepend long random string
.sp
5 \- Fake parameter
.sp
6 \- TAB as request spacer
.sp
7 \- Change the case of the URL
.sp
8 \- Use Windows directory separator (\e)
.RE
.PP
\fB\-findonly\fR
.RS 4
Only discover the HTTP(S) ports, do not perform a security scan\&. This will attempt to connect with HTTP or HTTPS, and report the Server header\&.
.RE
.PP
\fB\-Format\fR
.RS 4
Save the output file specified with \-o (\-output) option in this format\&. If not specified, the default will be taken from the file extension specified in the \-output option\&. Valid formats are:
.sp
csv \- a comma\-seperated list
.sp
htm \- an HTML report
.sp
txt \- a text report
.sp
xml \- an XML report
.RE
.PP
\fB\-host\fR
.RS 4
Host(s) to target\&. Can be an IP address, hostname or text file of hosts\&. A single dash (\-) maybe used for stdout\&. Can also parse nmap \-oG style output
.RE
.PP
\fB\-Help\fR
.RS 4
Display extended help information\&.
.RE
.PP
\fB\-id\fR
.RS 4
ID and password to use for host Basic host authentication\&. Format is "id:password"\&.
.RE
.PP
\fB\-list\-plugins\fR
.RS 4
Will list all plugins that Nikto can run against targets and then will exit without performing a scan\&. These can be tuned for a session using the \-plugins option\&.
.sp
The output format is:
.sp
Plugin
\fIname\fR
.sp
\ \&\fIfull name\fR
\-
\fIdescription\fR
.sp
\ \&Written by
\fIauthor\fR, Copyright (C)
\fIcopyright\fR
.RE
.PP
\fB\-mutate\fR
.RS 4
Specify mutation technique\&. A mutation will cause Nikto to combine tests or attempt to guess values\&. These techniques may cause a tremendous amount of tests to be launched against the target\&. Use the reference number to specify the type, multiple may be used:
.sp
1 \- Test all files with all root directories
.sp
2 \- Guess for password file names
.sp
3 \- Enumerate user names via Apache (/~user type requests)
.sp
4 \- Enumerate user names via cgiwrap (/cgi\-bin/cgiwrap/~user type requests)
.sp
5 \- Attempt to brute force sub\-domain names, assume that the host name is the parent domain
.sp
6 \- Attempt to guess directory names from the supplied dictionary file
.RE
.PP
\fB\-mutate\-options\fR
.RS 4
Provide extra information for mutates, e\&.g\&. a dictionary file
.RE
.PP
\fB\-nolookup\fR
.RS 4
Do not perform name lookups on IP addresses\&.
.RE
.PP
\fB\-nossl\fR
.RS 4
Do not use SSL to connect to the server\&.
.RE
.PP
\fB\-no404\fR
.RS 4
Disable 404 (file not found) checking\&. This will reduce the total number of requests made to the webserver and may be preferable when checking a server over a slow link, or an embedded device\&. This will generally lead to more false positives being discovered\&.
.RE
.PP
\fB\-output\fR
.RS 4
Write output to the file specified\&. The format used will be taken from the file extension\&. This can be over\-riden by using the \-Format option (e\&.g\&. to write text files with a different extenstion\&. Existing files will have new information appended\&.
.RE
.PP
\fB\-plugins\fR
.RS 4
Select which plugins will be run on the specified targets\&. A comma separated list should be provided which lists the names of the plugins\&. The names can be found by using \-list\-plugins\&.
.sp
There are two special entries: ALL, which specifies all plugins shall be run and NONE, which specifies no plugins shall be run\&. The default is ALL
.RE
.PP
\fB\-port\fR
.RS 4
TCP port(s) to target\&. To test more than one port on the same host, specify the list of ports in the \-p (\-port) option\&. Ports can be specified as a range (i\&.e\&., 80\-90), or as a comma\-delimited list, (i\&.e\&., 80,88,90)\&. If not specified, port 80 is used\&.
.RE
.PP
\fB\-Pause\fR
.RS 4
Seconds to delay between each test\&.
.RE
.PP
\fB\-root\fR
.RS 4
Prepend the value specified to the beginning of every request\&. This is useful to test applications or web servers which have all of their files under a certain directory\&.
.RE
.PP
\fB\-ssl\fR
.RS 4
Only test SSL on the ports specified\&. Using this option will dramatically speed up requests to HTTPS ports, since otherwise the HTTP request will have to timeout first\&.
.RE
.PP
\fB\-Single\fR
.RS 4
Perform a single request to a target server\&. Nikto will prompt for all options which can be specified, and then report the detailed output\&. See Chapter 5 for detailed information\&.
.RE
.PP
\fB\-timeout\fR
.RS 4
Seconds to wait before timing out a request\&. Default timeout is 10 seconds\&.
.RE
.PP
\fB\-Tuning\fR
.RS 4
Tuning options will control the test that Nikto will use against a target\&. By default, if any options are specified, only those tests will be performed\&. If the "x" option is used, it will reverse the logic and exclude only those tests\&. Use the reference number or letter to specify the type, multiple may be used:
.sp
0 \- File Upload
.sp
1 \- Interesting File / Seen in logs
.sp
2 \- Misconfiguration / Default File
.sp
3 \- Information Disclosure
.sp
4 \- Injection (XSS/Script/HTML)
.sp
5 \- Remote File Retrieval \- Inside Web Root
.sp
6 \- Denial of Service
.sp
7 \- Remote File Retrieval \- Server Wide
.sp
8 \- Command Execution / Remote Shell
.sp
9 \- SQL Injection
.sp
a \- Authentication Bypass
.sp
b \- Software Identification
.sp
c \- Remote Source Inclusion
.sp
x \- Reverse Tuning Options (i\&.e\&., include all except specified)
.sp
The given string will be parsed from left to right, any x characters will apply to all characters to the right of the character\&.
.RE
.PP
\fB\-useproxy\fR
.RS 4
Use the HTTP proxy defined in the configuration file\&.
.RE
.PP
\fB\-update\fR
.RS 4
Update the plugins and databases directly from cirt\&.net\&.
.RE
.PP
\fB\-Version\fR
.RS 4
Display the Nikto software, plugin and database versions\&.
.RE
.PP
\fB\-vhost\fR
.RS 4
Specify the Host header to be sent to the target\&.
.RE
.SH "FILES"
.PP
\fInikto\&.conf\fR
.RS 4
The Nikto configuration file\&. This sets Nikto\'s global options\&. Several nikto\&.conf files may exist and are parsed in the below order\&. As each configuration file is loaded is supersedes any previously set configuration:
.sp
.RS 4
\h'-04'\(bu\h'+03'System wide (e\&.g\&. /etc/nikto\&.conf)
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Home directory (e\&.g\&. $HOME/nikto\&.conf)
.RE
.sp
.RS 4
\h'-04'\(bu\h'+03'Current directory (e\&.g\&. \&./nikto\&.conf)
.RE
.RE
.PP
\fI${NIKTO_DIR}/plugins/db*\fR
.RS 4
db files are the databases that nikto uses to check for vulnerabilities and issues within the web server\&.
.RE
.PP
\fI${NIKTO_DIR}/plugins/*\&.plugin\fR
.RS 4
All nikto\'s plugins exist here\&. Nikto itself is just a wrapper script to manage CLI and pass through to the plugins\&.
.RE
.PP
\fI${NIKTO_DIR}/templates\fR
.RS 4
Contains the templates for nikto\'s output formats\&.
.RE
.SH "BUGS"
.PP
The current features are not supported:
.sp
.RS 4
\h'-04'\(bu\h'+03'SOCKS Proxies
.RE
.SH "AUTHORS"
.PP
Nikto was originally written and maintained by Sullo, CIRT, Inc\&. It is currently maintained by David Lodge\&. See the main documentation for other contributors\&.
.PP
All code is (C) CIRT, Inc\&., except LibWhisker which is (C) rfp\&.labs (wiretrip\&.net)\&. Other portions of code may be (C) as specified\&.
.SH "SEE ALSO"
.PP

\fINikto Homepage\fR\&[1]
.SH "NOTES"
.IP " 1." 4
Nikto Homepage
.RS 4
\%http://www.cirt.net/
.RE


docs/CHANGES.txt
01.06.2008 2.02
      - Added XML output thanks to the work of Jabra. XML format comes from templates (same as HTML). See the 'templates' dir for more info.
      - HTML reports changed by Jabra to remove some oddities and remove HTML from items
      - Fixed non-reporting of non-HTTP ports (or closed ports) when at least one port was HTTP.
      - Removed experimental knowledge base (KB) code, as XML output is more flexible for long-term scan tracking
      - Added unique identifiers to all tests from databases, and all tests created in code
      - Updated documentation
01.02.2008 nikto_core
      - Fixed improper parsing of long options (-update, etc.). Thanks to Frank Breedijk for figuring this out.
12.30.2007 db_servers
      - Removed as it is not used
12.19.2007 nikto_msgs.plugin
      - Add a boundary for regex on versions to cut down false positives
12.19.2007 niko_favicon.plugin
      - Added OSVDB ID
12.18.2007 niko_favicon.plugin
      - Fix false positive when favicon.ico doesn't exist
11.22.2007 Nikto 2.01 release
      - Fix anti ids encoding use. thanks to Francisco Amato
      - Fix virtual host usage if set via CLI. thanks Jon Hart
      - Fix Host header restoration when testing for IIS IP leak
      - Fix for plugindir & templatedir if EXECDIR is set in config.txt, thanks Shiraishi.M and Will Andrews for pointing it out.
      - Fix count of items--count now accurately reflects the number of items, not just number of vulns. thanks Frank Breedijk
      - Kick a few more things to KB that should be saved
      - Added SKIPIDS to config.txt to completely ignore some tests loaded from db_tests. Suggested by Christian Folini.
      - Enhanced rm_active_content to try to exclude the file/QUERYSTRING requested
      - Unset the auth header after guessing at it. Thanks Paul Woroshow for reporting the bug.
11.12.2007 nikto_headers.plugin
       - Fix internal IP address snarfing for IIS, thanks Frank Breedijk for pointing it out
11.10.2007 Nikto 2.00 release
       - Rewrite of nikto_httpoptions.plugin to read the Public header
       - Fixups to prevent namespace violations in nikto.pl and nikto_core.plugin
       - Add some normalizations to the -root option variable, suggested by Erik Cabetas
       - Added -Display with options for suppressing redirects & cookies from being included in output
       - Added -Tuning options to let users specify what they would like to test, or exclude certain categories
       - Added config.txt's NMAPOPTS, thanks Sean Lewis for the suggestion
       - All new HTML report
       - Bugfix: a found cookie would report for every port/server after it was found
       - Bugfix: all hosts scanned with all ports if hosts file used
       - Bugfix: all hosts scanned with port 80 despite what the user wanted
       - Bugfix: Reverse DNS inet_aton error fix, pointed out by Jason Peel @ Foundstone
       - Changed auth checking so it will test any directory found, not just /, and removed nikto_realms.plugin as a consequence
       - Changed scan_database.db format significantly (and name), (and all the code to deal with tests)
       - Completely new 404 engine which causes less false-positives (see docs)
       - Created dump_lw_hash instead of dump_request_hash & dump_result_hash
       - Implemented a knowledge base which (should) store all the gory details of scans... probably use this later ;)
       - Moved pre-defined variables from config.txt to variables.db so they can be automagically updated. Entries in config.txt are still read.
       - Removed %CFG, storing vars in %NIKTO instead
       - Removed -generic
       - Removed extraneous global vars
       - Removed load_realms, combined with load_variables
       - Replaced %CONFIG with %NIKTOCONFIG
       - Set MAX_WARN to trigger on any response code, skipping 404|403|401|400 to avoid common ones
       - Added -Single single request mode
       - Updates to use the RFP's LibWhisker 2.0
       - Added -Help to show extended help ouput, changed default help screen to be shorter. Suggested by Jericho.
       - Additional error checking on invalid reverse-dns (Paul Woroshow)
       - Cleaned up comment/line parsing routines in multiple places, from Erik Cabetas
       - Tightened some for loops with real values instead of guessing, from Erik Cabetas
       - Addded error message if no host is specified, from Erik Cabetas
       - Added more robust output file type checking (txt/htm/cvs), from Erik Cabetas
       - Added more debug statements regarding which CGI directories will be scanned, from Erik Cabatas
       - Bugfix: more 'half dead host' scanning issues resolved with Jericho. LW is much pickier now about calling http_close
       - Added error if -F specified without -o, from Erik Cabetas
       - Bugfix: server category match no longer matches partial strings, from Erik Cabetas
       - Bugfix: mis-pasted line, pointed to by Erik Cabetas
       - Send all errors to STDERR
       - Added -config option to specify a config file, thanks to Pavel Kankovsky
       - fixed regex issue on banner. thanks Alexander Ehlert for pointing it out
       - All other plugins updated for v2 changes
       - Added favicon.ico hash checking
       - ... gobs more

02.06.2004 nikto_core.plugin	1.21
	- Cleaned up comment/line parsing routines in multiple places, from Erik Cabetas
	- Tightened some for loops with real values instead of guessing, from from Erik Cabetas
	- Removed duplicate bit of code, from Erik Cabetas
	- Addded error message if no host is specified, from Erik Cabetas
	- Added more robust output file type checking (txt/htm/cvs), from Erik Cabetas
	- Added more debug statements regarding which CGI directories will be scanned, from Erik Cabatas

12.17.2003 
	nikto_core.plugin	1.20
       - Fixed BID links, thanks Richard Tortorella for the report.

10.27.2003 Nikto 1.32 release
	nikto_core.plugin	1.19
       - Removed unecessary 'use IO::Socket' call from resolve()

trunk/nikto.pl
#!/usr/bin/perl
use strict;

#VERSION,2.1.1
use Getopt::Long;
Getopt::Long::Configure('no_ignore_case');

###############################################################################
#                               Nikto                                         #
# --------------------------------------------------------------------------- #
#   $Id$                             #   
# --------------------------------------------------------------------------- #
###############################################################################
#  Copyright (C) 2004-2010 CIRT, Inc.
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; version 2
#  of the License only.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
# Contact Information:
#     Sullo (sullo@cirt.net)
#     http://cirt.net/
#######################################################################
# See the LICENSE.txt file for more information on the License Nikto is distributed under.
#
# This program is intended for use in an authorized manner only, and the author
# can not be held liable for anything done with this program, code, or items discovered
# with this program's use.
#######################################################################

# global var/definitions
use vars qw/$TEMPLATES %ERRSTRINGS %CLI %VARIABLES %TESTS $CONTENT/;
use vars qw/%NIKTO %REALMS %NIKTOCONFIG %request %result %COUNTERS/;
use vars qw/%db_extensions %FoF %UPDATES @DBFILE @BUILDITEMS/;
use vars qw/@RESULTS @PLUGINS @MARKS @REPORTS %CACHE/;

# setup
my $starttime        = localtime();
$NIKTO{'DIV'}        = "-" x 75;
$NIKTO{'version'}    = "2.1.1";
$NIKTO{'name'}       = "Nikto";
$NIKTO{'configfile'} = "/etc/nikto.conf";    ### Change this line if your setup is having trouble finding it

# read the --config option
{
    my %optcfg;
    Getopt::Long::Configure('pass_through', 'noauto_abbrev');
    GetOptions(\%optcfg, "config=s");
    Getopt::Long::Configure('nopass_through', 'auto_abbrev');
    if (defined $optcfg{'config'}) { $NIKTO{'configfile'} = $optcfg{'config'}; }
}

# Read the config files in order
my $error;
my $config_exists=0;
$error=load_config("$NIKTO{'configfile'}");
$config_exists=1 if ($error eq "");
# Guess home directory -- to support Windows
my $home="";
foreach my $var (split(/ /,"HOME USERPROFILE"))
{
   $home=$ENV{$var} if ($ENV{$var});
}
$error=load_config("$home/nikto.conf");
$config_exists=1 if ($error eq "");
$error=load_config("nikto.conf");
$config_exists=1 if ($error eq "");

if ($config_exists==0)
{
   die "- Could not find a valid nikto config file\n";
}

setup_dirs();
require "$NIKTOCONFIG{PLUGINDIR}/nikto_core.plugin";
nprint("T:$starttime: Starting", "d");
require "$NIKTOCONFIG{PLUGINDIR}/nikto_single.plugin";
require "$NIKTOCONFIG{PLUGINDIR}/LW2.pm";

# use LW2;                   ### Change this line to use a different installed version

($a, $b) = split(/\./, $LW2::VERSION);
die("- You must use LW2 2.4 or later\n") if ($a != 2 || $b < 4);

general_config();
load_databases();
load_databases('u');
nprint("- $NIKTO{'name'} v$NIKTO{'version'}");

LW2::http_init_request(\%request);
$request{'whisker'}->{'ssl_save_info'}              = 1;
$request{'whisker'}->{'lowercase_incoming_headers'} = 1;
$request{'whisker'}->{'timeout'}                    = $CLI{'timeout'} || 10;
if (defined $CLI{'evasion'}) { $request{'whisker'}->{'encode_anti_ids'} = $CLI{'evasion'}; }
$request{'User-Agent'} = $NIKTO{'useragent'};
$request{'whisker'}->{'retry'} = 0;
proxy_setup();

nprint($NIKTO{'DIV'});

# No targets - quit while we're ahead
if ($CLI{'host'} eq "") 
{ 
   nprint("+ ERROR: No host specified");
   usage(); 
}

$COUNTERS{hosts_total}=$COUNTERS{hosts_complete}=0;
load_plugins();

# Parse the supplied list of targets
my @MARKS=set_targets($CLI{'host'}, $CLI{'ports'}, $CLI{'ssl'}, $CLI{'root'});
# Now check each target is real and remove duplicates/fill in extra information
foreach my $mark (@MARKS)
{
   $mark->{test} = 1;
   # Try to resolve the host
   ($mark->{hostname}, $mark->{ip}, $mark->{display_name}) = resolve($mark->{ident});
   
   # Skip if we can't resolve the host - we'll error later
   if (!defined $mark->{ip})
   {
      $mark->{test} = 0;
      next;
   }

   # Check that the port is open
   my $open=port_check($mark->{hostname}, $mark->{ip}, $mark->{port});
   if (defined $CLI{'vhost'}) { $mark->{vhost} = $CLI{'vhost'} };
   if ($open == 0) {
      $mark->{test} = 0;
      next;
   }
   $mark->{ssl}=$open-1;
}

# Open reporting
report_head($CLI{'format'}, $CLI{'file'});

# Now we've done the precursor, do the scan
foreach my $mark (@MARKS) {
   next unless ($mark->{test});
   $COUNTERS{hosts_total}++;
   $mark->{start_time} = time();
   # These should just be passed in the hash - but it's a lot of work to move to a local $request
   $request{whisker}->{host} = $mark->{hostname} || $mark->{ip};
   if (defined $CLI{'vhost'})
   {
      $request{Host} = $CLI{'vhost'};
      $mark->{vhost} = $CLI{'vhost'};
   }
   $request{'whisker'}->{'port'}    = $mark->{port};
   $request{'whisker'}->{'ssl'}     = $mark->{ssl};
   $request{'whisker'}->{'version'} = $NIKTOCONFIG{DEFAULTHTTPVER};
   if (defined $NIKTOCONFIG{'STATIC-COOKIE'}) { $request{'Cookie'} = $NIKTOCONFIG{'STATIC-COOKIE'}; }
   $mark->{total_vulns}=0;
   $mark->{total_checks}=0;
   
   %FoF = ();
   
   $mark->{banner}=get_banner($mark);

   # put a signal trap so we can close down reports properly
   $SIG{'INT'} = sub
   {
      $mark->{end_time} = time();
      report_host_end($mark);
      report_close($mark);
      exit(1);
   };
   
   if ($CLI{'findonly'})
   {
      my $protocol="http";
      if ($mark->{ssl}) { $protocol .= "s"; }
      if ($mark->{banner} eq "")
      {
         $mark->{banner} = "(no identification possible)";
      }
      nprint("+ Server: $protocol://$mark->{display_name}:$mark->{port}\t$mark->{banner}");
   }
   else
   {
      dump_target_info($mark);
      report_host_start($mark);
      set_scan_items($mark);
      unless (defined $CLI{'nofof'}) { map_codes() };
      run_plugins($mark);
   }
   $mark->{end_time} = time();
   my $time=date_disp($mark->{end_time});
   my $elapsed=$mark->{end_time}-$mark->{start_time};
   nprint("+ $mark->{total_checks} items checked: $mark->{total_vulns} item(s) reported on remote host");
   nprint("+ End Time:           $time ($elapsed seconds)");
   nprint($NIKTO{'DIV'});
   
   $COUNTERS{hosts_completed}++;
   report_host_end($mark);
}
report_close();
   
nprint("+ $COUNTERS{hosts_total} host(s) tested");
nprint("+ $NIKTO{'totalrequests'} requests made","v");
send_updates();
nprint("T:" . localtime() . ": Ending", "d");

exit;

#################################################################################
####                Most code is now in nikto_core.plugin                    ####
#################################################################################
# load config file
# error=load_config(FILENAME)
sub load_config
{
   my $configfile=$_[0];

   open(CONF, "<$configfile") || return "+ ERROR: Unable to open config file '$configfile'";
   my @CONFILE = <CONF>;
   close(CONF);

   foreach my $line (@CONFILE)
   {
      $line =~ s/\#.*$//;
      chomp($line);
      $line =~ s/\s+$//;
      $line =~ s/^\s+//;
      next if ($line eq "");
      my @temp = split(/=/, $line, 2);
      if ($temp[0] ne "") { $NIKTOCONFIG{ $temp[0] } = $temp[1]; }
   }

   # add CONFIG{CLIOPTS} to ARGV if defined...
   if (defined $NIKTOCONFIG{CLIOPTS})
   {
      my @t = split(/ /, $NIKTOCONFIG{CLIOPTS});
      foreach my $c (@t) { push(@ARGV, $c); }
   }

   # Check for necessary config items
   check_config_defined("CHECKMETHODS", "HEAD");

   return "";
}
#################################################################################
# find plugins directory
sub setup_dirs
{
   my $CURRENTDIR = $0;
   chomp($CURRENTDIR);
   $CURRENTDIR =~ s#[\\/]nikto.pl$##;

   # First assume we get it from NIKTOCONFIG
   unless (defined $NIKTOCONFIG{EXECDIR})
   {
      if (-d "$ENV{PWD}/plugins")
      {
         $NIKTOCONFIG{EXECDIR}=$ENV{PWD};
      }
      elsif (-d "$CURRENTDIR/plugins")
      {
         $NIKTOCONFIG{EXECDIR}=$CURRENTDIR;
      }
      elsif (-d "./plugins")
      {
         $NIKTOCONFIG{EXECDIR}=$CURRENTDIR;
      }
      else
      {
         print STDERR "Could not work out the nikto EXECDIR, try setting it in niktorc";
         exit;
      }
   }
   unless (defined $NIKTOCONFIG{PLUGINDIR})
   {
      $NIKTOCONFIG{PLUGINDIR}="$NIKTOCONFIG{EXECDIR}/plugins";
   }
   unless (defined $NIKTOCONFIG{TEMPLATEDIR})
   {
      $NIKTOCONFIG{TEMPLATEDIR}="$NIKTOCONFIG{EXECDIR}/templates";
   }
   unless (defined $NIKTOCONFIG{DOCUMENTDIR})
   {
      $NIKTOCONFIG{DOCUMENTDIR}="$NIKTOCONFIG{EXECDIR}/docs";
   }
   return;
}

######################################################################
## check_config_defined(item, default)
## Checks whether config has been set, warns and sets to a default
sub check_config_defined
{
   my $item=$_[0];
   my $default=$_[1];
 
   if (!defined $NIKTOCONFIG{$item})
   {
      print STDERR "- Warning: $item is not defined in Nikto configuration, setting to \"$default\"\n"; 
      $NIKTOCONFIG{$item}=$default;
   }
}
